src/atomic_file/atomic.rs

```diff
@@ -0,0 +1,170 @@
+    /// has not been created yet.
+    pub fn open(&self) -> Result<Option<File>> {
+        if self.version != 0 {
+            Ok(Some(File::open(&self.path)?))
+        } else {
+            Ok(None)
+        }
+    }
+
+    pub fn read_to_string(&self) -> Result<String> {
+        match self.open() {
+            Ok(None) => Err(Error::new(ErrorKind::NotFound, "File not found")),
+            Ok(Some(mut file)) => {
+                let mut buff = String::new();
+                file.read_to_string(&mut buff)?;
+                Ok(buff)
+            }
+            Err(e) => Err(e),
+        }
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct AtomicFile {
+    directory: PathBuf,
+    prefix: String,
+}
+
+fn parse_version(filename: &std::ffi::OsStr, prefix: &str) -> Option<usize> {
+    let filename = filename.to_str()?;
+    if !filename.starts_with(prefix) {
+        return None;
+    }
+    filename[prefix.len()..].parse().ok()
+}
+
+impl AtomicFile {
+    pub fn new(path: impl Into<PathBuf>) -> Result<Self> {
+        let directory = path.into();
+        std::fs::create_dir_all(&directory)?;
+        let filename: &str = match directory.file_name() {
+            Some(name) => name.to_str().unwrap(),
+            None => {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::InvalidInput,
+                    "`path` must specify a directory name",
+                ));
+            }
+        };
+        let prefix = format!("{}.", filename);
+        Ok(Self { directory, prefix })
+    }
+
+    fn latest_version(&self) -> Result<usize> {
+        let mut max_version = 0;
+        for entry in fs::read_dir(&self.directory)? {
+            if let Some(version) =
+                parse_version(&entry?.file_name(), &self.prefix)
+            {
+                max_version = std::cmp::max(max_version, version);
+            }
+        }
+        Ok(max_version)
+    }
+
+    fn path(&self, version: usize) -> PathBuf {
+        self.directory
+            .join(format!("{}{version}", self.prefix))
+    }
+
+    pub fn load(&self) -> Result<ReadOnlyFile> {
+        let version = self.latest_version()?;
+        let path = self.path(version);
+        Ok(ReadOnlyFile { version, path })
+    }
+
+    pub fn make_temp(&self) -> Result<TmpFile> {
+        TmpFile::create_in(&self.directory)
+    }
+
+    /// Replace the contents of the file with the contents of `new` if the
+    /// latest version is the same as `current`.
+    ///
+    /// # Errors
+    /// If `io::ErrorKind::AlreadyExists` is returned, it means that the latest
+    /// version was not the same as `current` and the operation must be retried
+    /// with a fresher version of the file. Any other I/O error is forwarded as
+    /// well.
+    pub fn compare_and_swap(
+        &self,
+        current: &ReadOnlyFile,
+        new: TmpFile,
+    ) -> Result<()> {
+        let new_path = self.path(current.version + 1);
```

Comment:
It depends of the architecture but on 32 bits machine 4_294_967_295
and 18_446_744_073_709_551_615 on 64 bits machines.
On release mode it will just overflow and restart at 0
Concerning other filename convention, the current implementation used the version number to load the last file. So if it's not the case how to determine the latestt file?

Reply:
> It depends of the architecture but on 32 bits machine 4_294_967_295
and 18_446_744_073_709_551_615 on 64 bits machines.

Sounds good, especially this:
> On release mode it will just overflow and restart at 0

Reply:
> Concerning other filename convention, the current implementation used the version number to load the last file. So if it's not the case how to determine the latest file?

This is correct and not needed to be changed.

We only need to introduce additional suffix for the filename — _device id_.
It should be:
* Either computed deterministically from the hardware used (fingerprint)
* Or generated once and put somewhere in user settings

Reading:
* Find latest version using incremental numbers (same as now).
* If several versions with same number are found, take the version with local _device id_ and log warning about conflicted versions. These conflicts can come only from remote devices, locally we can't produce conflicts.